rasa/shared/core/domain.py

```diff
@@ -1068,15 +1068,90 @@ def get_active_states(self, tracker: "DialogueStateTracker") -> State:
         }
         return self._clean_state(state)
 
+    @staticmethod
+    def _remove_rule_only_features(
+        state: State, tracker: "DialogueStateTracker"
+    ) -> None:
+        # remove only rule slots
+        for slot in tracker.rule_only_slots:
+            if state.get(rasa.shared.core.constants.SLOTS, {}).get(slot):
+                del state[rasa.shared.core.constants.SLOTS][slot]
+        # remove active loop feature if it is only rule loop
+        if (
+            state.get(rasa.shared.core.constants.ACTIVE_LOOP, {}).get(
+                rasa.shared.core.constants.LOOP_NAME
+            )
+            in tracker.rule_only_loops
+        ):
+            del state[rasa.shared.core.constants.ACTIVE_LOOP]
+
+    @staticmethod
+    def _substitute_rule_only_user_input(state: State, last_ml_state: State) -> None:
+        if not rasa.shared.core.trackers.is_prev_action_listen_in_state(state):
+            if not last_ml_state.get(rasa.shared.core.constants.USER) and state.get(
+                rasa.shared.core.constants.USER
+            ):
+                del state[rasa.shared.core.constants.USER]
+            elif last_ml_state.get(rasa.shared.core.constants.USER):
+                state[rasa.shared.core.constants.USER] = last_ml_state[
+                    rasa.shared.core.constants.USER
+                ]
+
     def states_for_tracker_history(
-        self, tracker: "DialogueStateTracker"
+        self, tracker: "DialogueStateTracker", ignore_rule_only_turns: bool = False
     ) -> List[State]:
-        """Array of states for each state of the trackers history."""
-        return [
-            self.get_active_states(tr) for tr in tracker.generate_all_prior_trackers()
-        ]
+        """Array of states for each state of the trackers history.
+
+        Args:
+            tracker: Instance of `DialogueStateTracker` to featurize.
+            ignore_rule_only_turns: If True ignore dialogue turns that are present
+                only in rules.
+
+        Return:
+            A list of states.
+        """
+        states = []
+        last_ml_action_sub_state = None
+        last_action_is_ml_action = True
+        for tr in tracker.generate_all_prior_trackers():
+            if ignore_rule_only_turns:
+                # remember previous ml action based on the last non hidden turn
+                # we need this to override previous action in the ml state
+                if last_action_is_ml_action:
+                    last_ml_action_sub_state = self._get_prev_action_sub_state(tr)
+                last_action_is_ml_action = not tr.hide_rule_turn
+
+                if tr.hide_rule_turn:
+                    continue
+
+            state = self.get_active_states(tr)
+
+            if ignore_rule_only_turns:
```

Comment:
can we concat this with the previous `if ignore_rule_only_turns`?

Reply:
what do you mean? state need to be calculated in between

Reply:
can't we calculate the state before the previous `if`?

Reply:
then it is unnecessary state calculation. Previous if contains `continue`